site/app/controllers/MiscController.php

```diff
@@ -58,4 +72,197 @@ private function display_file() {
             }
         }
     }
+
+    private function downloadFile() {
+        //Additional security
+        if (!($this->core->getUser()->accessGrading())) {
+            return false;
+        }
+        if (!file_exists($_REQUEST['path'])) {
+            throw new \InvalidArgumentException("File does not exist");
+        }
+        foreach (explode(DIRECTORY_SEPARATOR, $_REQUEST['path']) as $part) {
+            if ($part == ".." || $part == ".") {
+                throw new \InvalidArgumentException("Cannot have a part of the path just be dots");
+            }
+        }
+        $this->core->getOutput()->useHeader(false);
+        $this->core->getOutput()->useFooter(false);
+        $file_url = $_REQUEST['path'];
+        header('Content-Type: application/octet-stream');
+        header("Content-Transfer-Encoding: Binary"); 
+        header("Content-disposition: attachment; filename=\"" . basename($file_url) . "\""); 
+        readfile($file_url);
+    }
+
+    private function downloadZip() {
+        //Additional security
+        if (!($this->core->getUser()->accessGrading())) {
+            return false;
+        }
+        $zip_file_name = $_REQUEST['gradeable_id'] . "_" . $_REQUEST['user_id'] . "_" . date("m-d-Y") . ".zip";
+        $this->core->getOutput()->useHeader(false);
+        $this->core->getOutput()->useFooter(false);
+        $temp_dir = "/tmp";
+        //makes a random zip file name on the server
+        $temp_name = md5(uniqid($this->core->getUser()->getId(), true));
```

Comment:
If there's not a particular reason to take the MD5 hash here, this seems unnecessary, especially given that you're already setting passing `more_entropy = true` to `uniqid()`.